sources/TileDB.CSharp/Subarray.cs

```diff
@@ -0,0 +1,365 @@
+            return result;
+        }
+
+        /// <summary>
+        /// Gets the number of ranges for a given dimension name.
+        /// </summary>
+        /// <param name="dimensionName">The dimension's name.</param>
+        public ulong GetRangeCount(string dimensionName)
+        {
+            using var ctxHandle = _ctx.Handle.Acquire();
+            using var handle = _handle.Acquire();
+            using var ms_dimensionName = new MarshaledString(dimensionName);
+            ulong result;
+            _ctx.handle_error(Methods.tiledb_subarray_get_range_num_from_name(ctxHandle, handle, ms_dimensionName, &result));
+            return result;
+        }
+
+        /// <summary>
+        /// Gets the range for a given dimension index and range index.
+        /// </summary>
+        /// <typeparam name="T">The dimension's type.</typeparam>
+        /// <param name="dimensionIndex">The dimension's index.</param>
+        /// <param name="rangeIndex">The range's index.</param>
+        /// <returns>The dimension's start and end values.</returns>
+        /// <exception cref="NotSupportedException"><typeparamref name="T"/> is not supported.</exception>
+        public (T Start, T End) GetRange<T>(uint dimensionIndex, uint rangeIndex) where T : struct
+        {
+            _ = EnumUtil.TypeToDataType(typeof(T));
+            using var ctxHandle = _ctx.Handle.Acquire();
+            using var handle = _handle.Acquire();
+            void* startPtr, endPtr, stridePtr_unused;
+            _ctx.handle_error(Methods.tiledb_subarray_get_range(ctxHandle, handle, dimensionIndex, rangeIndex, &startPtr, &endPtr, &stridePtr_unused));
+            return (Unsafe.ReadUnaligned<T>(startPtr), Unsafe.ReadUnaligned<T>(endPtr));
+        }
+
+        /// <summary>
+        /// Gets the range for a given dimension name and range index.
+        /// </summary>
+        /// <typeparam name="T">The dimension's type.</typeparam>
+        /// <param name="dimensionName">The dimension's name.</param>
+        /// <param name="rangeIndex">The range's index.</param>
+        /// <returns>The dimension's start and end values.</returns>
+        /// <exception cref="NotSupportedException"><typeparamref name="T"/> is not supported.</exception>
+        public (T Start, T End) GetRange<T>(string dimensionName, uint rangeIndex) where T : struct
+        {
+            _ = EnumUtil.TypeToDataType(typeof(T));
+            using var ctxHandle = _ctx.Handle.Acquire();
+            using var handle = _handle.Acquire();
+            using var ms_dimensionName = new MarshaledString(dimensionName);
+            void* startPtr, endPtr, stridePtr_unused;
+            _ctx.handle_error(Methods.tiledb_subarray_get_range_from_name(ctxHandle, handle, ms_dimensionName, rangeIndex, &startPtr, &endPtr, &stridePtr_unused));
+            return (Unsafe.ReadUnaligned<T>(startPtr), Unsafe.ReadUnaligned<T>(endPtr));
+        }
```

Comment:
The equivalent `Query` methods did not validate types and neither do these; we could make them to now or in a future PR.